src/Analyzer.Reports/ReportsHelper.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Azure.Templates.Analyzer.Types;
+using Microsoft.CodeAnalysis.Sarif;
+
+namespace Microsoft.Azure.Templates.Analyzer.Reports
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public class ReportsHelper
+    {
+        /// <summary>
+        /// Get all distinct results in an evaluation organized by source location file
+        /// </summary>
+        /// <param name="evaluations">Evaluations to get results for</param>
+        /// <param name="filesToSkip">Files to not include results from</param>
+        /// <returns></returns>
+        public static Dictionary<string, List<(IEvaluation, IList<IResult>)>> GetResultsByFile(
+            IEnumerable<Types.IEvaluation> evaluations,
+            IEnumerable<string> filesToSkip)
+        {
+            return GetResultsByFile(evaluations, filesToSkip, out int _);
+        }
+
+        /// <summary>
+        /// Get all distinct results in an evaluation organized by source location file
+        /// </summary>
+        /// <param name="evaluations">Evaluations to get results for</param>
+        /// <param name="filesToSkip">Files to not include results from</param>
+        /// <param name="passedEvaluations">Out parameter that gives number of passed evaluations</param>
+        /// <returns></returns>
+        public static Dictionary<string, List<(IEvaluation, IList<IResult>)>> GetResultsByFile(
+            IEnumerable<Types.IEvaluation> evaluations,
+            IEnumerable<string> filesToSkip,
+            out int passedEvaluations)
+        {
+            var resultsByFile = new Dictionary<string, List<(IEvaluation, IList<IResult>)>>();
+            passedEvaluations = 0;
+
+            foreach (var evaluation in evaluations)
+            {
+                if (!evaluation.Passed)
+                {
+                    (var actualFile, var failedResults) = GetResultsByFileInternal(evaluation);
+
+                    // a file's results may have already been output if analyzing directory
+                    if (filesToSkip.Contains(actualFile))
```

Comment:
Instead of skipping evaluations that are from the same file, let's just not analyze files that have already been analyzed.  This will make execution way more performant in directories that have lots of Bicep modules in use.  The only scenario we need to be cautious of is if parameters are provided when analyzing, so we should reevaluate the same file if different parameters are passed.

It should be as simple as
1. Track file identifiers analyzed in `TemplateAnalyzer.AnalyzeTemplate` (with the parameters used in analysis), and return the evaluations already found if it's been scanned before.
2. Track the same thing in the CLI, but simply don't call `TemplateAnalyzer.AnalyzeTemplate` for a file that's already been analyzed.

(2) should be done to optimize the most common use case (executing from the command line).  (1) should be done just to benefit other scenarios where Analyzer.Core is used directly in another project outside BPA.  This isn't as high of a priority.

Reply:
I initially skipped this optimization because I wasn't sure moving forward if we'd really want it or not--though if our goal is not to increase the granularity of data in our results (i.e. preserving more of the module information) then I suppose we could go ahead with this.

Reply:
OK this currently is actually not feasible because we do not have the "template reference tree" metadata to be able to determine the parent/child of any given analyzed template. This prevents us from filtering all the evaluations to the ones specific to each referenced module when tracking this data (i.e. all the evals that it directly or indirectly references through its own modules).

I think I could technically rebuild this relational data from the source map but it wouldn't be pretty or ideal. This is also a related gap affecting the ARM-template-as-bicep-module issue as well as the other comment on this PR regarding the BicepTemplateProcessor. I propose to create a bug for this work and I can do the work once we have the data available to us.

Reply:
OK--even with the new source data I hit a snag here. We need to be able to bucket each top-level evaluation by the source file it was evaluated in (specifically the resource that it evaluated against). However, I am seeing some evaluations that do not contain any results, and consequently no SourceLocations to pull the source file from. Using the TemplateIdentifier is inaccurate for results in modules. So, this means I don't know what file the evaluations are actually from if there's modules in the entrypoint.

Considering tabling this for a later PR considering it's an optimization and the cost is unknown but growing still.

Reply:
#308 to track. Not sure we have an issue yet specifically with the evaluations with no results. I want to investigate whether this only happens with passing evaluations, as in that case we could duplicate the evaluation as it would apply to each dependent template.